### DISCUSSION TITLE
bad crash application scroll item property - azure

### DISCUSSION THREAD

**Discussion by Paolo-lab** [iTwin/itwinjs-core#5791] [Q&A] [answered]
Hello good evening, my end customer noticed a problem with my application located on Azure, I state that I don't have this problem locally, but it only happens on the application located in azure.

when scrolling through the properties of an element the program often crashes, unfortunately it's not code written by me but it's an internal problem I don't know if it's due to azure, maybe undersized or what else I'm attaching two images of the error ..
The customer asks me for explanations, but frankly I don't know what to answer

Thank you, really thank you
<img width="677" alt="azure_error1" src="https://github.com/iTwin/itwinjs-core/assets/57446652/0752dceb-fa4d-434e-b574-24c25a3c5d33">
<img width="522" alt="azure_error2" src="https://github.com/iTwin/itwinjs-core/assets/57446652/24ad5b03-2254-439f-8f08-98b15ca0b9c1">

**Comment by aruniverse** [comment]
Could you provide some more information??
To confirm, this is your own application?
What version of iTwin.js are you using? Share your package.json.
Please share a better call stack

**Comment by Paolo-lab** [comment]
hi this is my package.json 
{
  "name": "web-viewer-app",
  "version": "0.1.0",
  "private": true,
  "proxy": "https://localhost:44388",
  "dependencies": {
    "@bentley/icons-generic": "^1.0.34",
    "@bentley/react-scripts": "^4.0.7",
    "@itwin/appui-abstract": "^3.7.2",
    "@itwin/appui-layout-react": "^3.7.2",
    "@itwin/appui-react": "^3.7.2",
    "@itwin/browser-authorization": "^0.5.1",
    "@itwin/components-react": "^3.7.2",
    "@itwin/core-bentley": "^3.7.2",
    "@itwin/core-common": "^3.7.2",
    "@itwin/core-frontend": "^3.7.2",
    "@itwin/core-geometry": "^3.7.2",
    "@itwin/core-i18n": "^3.7.2",
    "@itwin/core-markup": "^3.7.2",
    "@itwin/core-orbitgt": "^3.7.2",
    "@itwin/core-quantity": "^3.7.2",
    "@itwin/core-react": "^3.7.2",
    "@itwin/core-telemetry": "^3.7.2",
    "@itwin/imodel-components-react": "^3.7.2",
    "@itwin/imodels-access-frontend": "^1.0.2",
    "@itwin/imodels-client-management": "^1.0.1",
    "@itwin/itwinui-layouts-css": "^0.2.0",
    "@itwin/itwinui-layouts-react": "^0.2.0",
    "@itwin/itwinui-react": "^1.48.2",
    "@itwin/measure-tools-react": "^0.7.3",
    "@itwin/presentation-common": "^3.7.2",
    "@itwin/presentation-frontend": "^3.7.2",
    "@itwin/property-grid-react": "^0.6.0",
    "@itwin/reality-data-client": "^0.7.2",
    "@itwin/tree-widget-react": "^0.4.7",
    "@itwin/web-viewer-react": "^3.1.4",
    "@itwin/webgl-compatibility": "^3.7.2",
    "@types/history": "^4.7.11",
    "@types/lodash.isempty": "^4.4.7",
    "@types/node": "^14.18.42",
    "@types/react": "^17.0.58",
    "@types/react-dom": "^17.0.19",
    "@types/react-router-dom": "^5.1.2",
    "axios": "^1.1.3",
    "bootstrap": "^5.1.3",
    "chart.js": "^3.7.1",
    "google-map-react": "^2.2.1",
    "history": "^4.10.1",
    "js-file-download": "^0.4.12",
    "react": "^17.0.2",
    "react-bootstrap": "^2.7.4",
    "react-chartjs-2": "^4.3.1",
    "react-dom": "^17.0.2",
    "react-redux": "^7.2.9",
    "react-router-dom": "^5.1.2",
    "read-excel-file": "^5.2.26",
    "redux": "^4.2.1",
    "typescript": "~4.4.0"
  },
  "scripts": {
    "start": "react-scripts --max_old_space_size=4096 start",
    "build": "react-scripts --max_old_space_size=4096 build",
    "test": "react-scripts --max_old_space_size=4096 test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": [
    "last 4 chrome version",
    "last 4 firefox version",
    "last 4 safari version",
    "last 4 ios version",
    "last 4 ChromeAndroid version",
    "last 4 edge version",
    "not dead",
    "not <0.2%"
  ],
  "devDependencies": {
    "@types/google-map-react": "^2.1.7"
  }
}

**Comment by Paolo-lab** [comment]
Hello good morning, last night I did not sleep a wink for this problem that makes my heart ache.
when I browse the properties it seems to me that it goes into error when it encounters some external path like the following

https://connect-projectshareweb.bentley.com/a2a4966d-xxxxx-4c15-b5ee-xxxxx/312f4de4-3d0a-xxxxx-aca3-3909bb63a850

or

pw://xxxxxx.xxxxx.corp.local:ITF_PWDS02/Documents/D{3771c154-970f-4435-962f-xxxxx}

I don't know if it can serve as an indication

thanks again for the support

**Comment by grigasp** [comment]
From the call stack it's pretty clear that `TypeConverterManager.getConverter` returns `undefined` [here](https://github.com/iTwin/itwinjs-core/blob/475e9ebb08ed6c50b6216a12001ac816bb0d6252/ui/components-react/src/components-react/properties/renderers/value/PrimitivePropertyValueRenderer.tsx#L43). However, that should not be possible based on [API declaration](https://github.com/iTwin/itwinjs-core/blob/475e9ebb08ed6c50b6216a12001ac816bb0d6252/ui/components-react/src/components-react/converters/TypeConverterManager.ts#L50) - either it [returns a registered converter](https://github.com/iTwin/itwinjs-core/blob/475e9ebb08ed6c50b6216a12001ac816bb0d6252/ui/components-react/src/components-react/converters/TypeConverterManager.ts#L54) that [can't be `undefined`](https://github.com/iTwin/itwinjs-core/blob/475e9ebb08ed6c50b6216a12001ac816bb0d6252/ui/components-react/src/components-react/converters/TypeConverterManager.ts#L34C4-L35) or the [default `StringTypeConverter`](https://github.com/iTwin/itwinjs-core/blob/475e9ebb08ed6c50b6216a12001ac816bb0d6252/ui/components-react/src/components-react/converters/TypeConverterManager.ts#L56), which should always get registered statically (see [here](https://github.com/iTwin/itwinjs-core/blob/475e9ebb08ed6c50b6216a12001ac816bb0d6252/ui/components-react/src/components-react/converters/TypeConverterManager.ts#L36-L38) and [here](https://github.com/iTwin/itwinjs-core/blob/475e9ebb08ed6c50b6216a12001ac816bb0d6252/ui/components-react/src/components-react/converters/StringTypeConverter.ts#L130-L132)).

I would say the most likely problem is that for some reason the `StringTypeConverter` doesn't get registered. The only way I see it could happen is if you use direct module imports rather than the barrel file (e.g. `import { UiComponents } from "@itwin/components-react/lib/cjs/components-react/UiComponents"` instead of `import { UiComponents } from "@itwin/components-react"`). Can you check that you always import from the barrel?

Also tagging @iTwin/itwinjs-core-ui team in case they have any more ideas.

**Comment by Paolo-lab** [comment]
Hi good morning ,

UiComponents I don't use it in my application by searching it is present in the asset-manifest.json file in the build folder
"locales/en/UiComponents.json": "/locales/en/UiComponents.json", and in build\static\js files

I thought it might be a localization problem but the console does not fail

<img width="899" alt="azure_error3" src="https://github.com/iTwin/itwinjs-core/assets/57446652/ca4b3599-6d4f-4b51-9670-72b97ea6e028">

**Comment by Paolo-lab** [comment]
@iTwin/itwinjs-core-ui

HI

I'm still trying to figure out why everything works locally and doesn't on azure
I found "convertPrimitiveRecordToString" which are different from each other

that is, this is what I find locally, I attach the image

<img width="846" alt="Version Local" src="https://github.com/iTwin/itwinjs-core/assets/57446652/6c38460e-0a7e-4396-800f-32f4e6af3ef4">


and this is what I find on Azure I attach image

<img width="691" alt="Version Azure" src="https://github.com/iTwin/itwinjs-core/assets/57446652/59d30712-fca0-422e-80b1-a7bbf234a688">


I don't think this thing is normal here ..

**Reply by grigasp** [reply]
The first image is javascript that results from transpiling typescript in the second image. From your screenshot it looks like you're on debugger in the second image - it could be that your debugger picks up source maps and maps javascript to typescript for you.

**Reply by Paolo-lab** [reply]
i would come back to this answer , does typescript transpiling add checksum ? like if null you do one thing else you do another thing ?

in the convertPrimitiveRecordToString the two versions the one on my compurer and the one on azure (let's say) disagree in the code on my compure he makes these considerations that on azure there are not ..

what can depend on the version of the package ?

I'm sorry but I don't know which fish to catch..

**Reply by grigasp** [reply]
Not sure I understand what you're asking. 

I would assume locally you're running the app through webpack devserver and for deployed case you:
1. build the source
2. deploy build output
3. run it

Can you try to download the build output that you're running on Azure and run it locally so see how it behaves there?

**Reply by Paolo-lab** [reply]
- Not sure I understand what you're asking.

I meant that the two versions seem too discordant with each other
the one on my computer that checks if the name is null or not, while the one that goes into error on azure, does not make these considerations

to launch my application locally I do "npm start"

to deply I do "npm run build" then I compress the build folder and through "scm" I copy it to azure, that's all I don't do

**Reply by grigasp** [reply]
I will assume `npm start` runs webpack devserver :)

You can run webpack build output locally to see if it behaves the same as when deployed on Azure. To do that:
```
npm install -g http-server
http-server {path_to_your_build_output}
```
This should start an HTTP server on your localhost.

**Reply by Paolo-lab** [reply]
hello, unfortunately not ...
this is my start
"start": "react-scripts --max_old_space_size=4096 start

**Reply by aruniverse** [reply]
your start command uses react-scripts which spawns a webpack dev server

run the following:
```
npm run build
npm install -g http-server
http-server build
```

that will serve a local copy of your production build used in your deployment
see if you can reproduce the issue there

**Comment by grigasp** [comment]
There's a possibility that your app uses two copies of `@itwin/components-react` package. Can you share your package lockfile for us to check that?

**Reply by Paolo-lab** [reply]
hello, I think at this point is the only solution, where I can send you the package

please, these are confidential things, I trust you ...

**Comment by Paolo-lab** [comment]
sorry when you were talking about the package I thought of the whole application that's why I asked you for discretion, sorry.

I am sending you the package-lock.json , and I think you are right

I have different version for example
I use @itwin/measure-tools-react vers. 0.7.3 which carries @itwin/components-react version 3.0.0

I use @itwin/property-grid-react vers. 0.6.0 which carries @itwin/components-react version 3.0.0

I use @itwin/tree-widget-react" vers. 0.4.7 which carries the 3.0.0 version of @itwin/components-react

I use @itwin/viewer-react ver. 3.4.1 which carries @itwin/components-react version 3.2.0

I also attach an image , which versions of these packages should I use to standardize @itwin/components-react to version 3.7.2 ?

I attach the zip of the file


<img width="392" alt="itwin-components-react" src="https://github.com/iTwin/itwinjs-core/assets/57446652/38847c5c-e930-432a-99c6-5583a37d2e39">


[package-lock.zip](https://github.com/iTwin/itwinjs-core/files/12193854/package-lock.zip)

**Reply by grigasp** [reply]
I don't see anything wrong in the lockfile - all `@itwin/components-react` dependencies resolve to the same `3.7.2` version.

**Reply by raplemie** [reply]
Other possibility is that you are building using some packages with common js and other with ES modules, @aruniverse is there a way to validate this with the react-scripts ? (Ensure that only one build is used for a particular package ?)

**Reply by aruniverse** [reply]
I would hope that's not the case here. Best way to analyze that is to build with webpack stats and analyze the bundle

**Comment by Paolo-lab** [comment]
Hi guys, any other ideas? Who can I contact at bentley-itwin to resolve this issue.

I have to open a ticket to call someone.

Today is the release date everything seemed to go smoothly and instead ..

**Comment by Paolo-lab** [comment]
hello good morning, yesterday among the test lights I tried to install

npm install webpack webpack-cli webpack-dev-server --save-dev
npm install @types/webpack-dev-server --save-dev

I would like to go back to the original configuration, but nothing works anymore

I uninstalled webpack packages webpack-cli webpack-dev-server --save-dev and types @types/webpack-dev-server --save-dev

I deleted the node_modules folder and reinstalled everything

I also tried to double the project and reinstall all packages again


but i keep getting this error on start this is the command
"start": "react-scripts --max_old_space_size=4096 start",

C:\Users\ppetracchiola\source\repos\SviluppoBentley\reactappNew\node_modules\@bentley\react-scripts\scripts\start.js:19
   throw err;
   ^

TypeError: Cannot read properties of undefined (reading 'createSnapshot')
     at C:\Users\ppetracchiola\source\repos\SviluppoBentley\reactappNew\node_modules\html-webpack-plugin\lib\webpack5\file-watcher-api.js:13:36
     at new Promise (<anonymous>)
     at Object.createSnapshot (C:\Users\ppetracchiola\source\repos\SviluppoBentley\reactappNew\node_modules\html-webpack-plugin\lib\webpack5\file-watcher-api.js:12:10)
     at C:\Users\ppetracchiola\source\repos\SviluppoBentley\reactappNew\node_modules\html-webpack-plugin\lib\cached-child-compiler.js:219:35
     at processTicksAndRejections (node:internal/process/task_queues:96:5)

what can I do because now all these problems that I have never had since I started the development of bentley?

**Reply by Paolo-lab** [reply]
I'm closing this help request and opening a separate one from the azure problem, sorry but I'm getting a bit confused

**Comment by Paolo-lab** [comment, marked as answer]
hello all good morning, after several attempts I upgraded from the @bentley/react-scripts^4.0.7 version to the @bentley/react-scripts^5.0.1 version

and everything started working fine both locally and on Azure.

Thank you for the tip on using http-server and for the support you have given me

Thank you
